README.md

```diff
@@ -1,44 +1,284 @@
+It compiles seccomp filters expressed as Rust code, into BPF filters, ready to
+be loaded into the kernel. This is the seccompiler **backend.**
 
-## Usage
+The process of translating JSON filters into BPF goes through an extra step of
+deserialization and validation (the **JSON frontend**), before reaching the
+same backend for BPF codegen.
 
-TODO: This section describes how the crate is used.
+The Rust representation is therefore also an Intermediate Representation (IR)
+of the JSON filter.
+This modular implementation allows for extendability in regards to file
+formats. All that is needed is a compatible front-end.
 
-Some questions that might help in writing this section:
-- What traits do users need to implement?
-- Does the crate have any default/optional features? What is each feature
-  doing?
-- Is this crate used by other rust-vmm components? If yes, how?
+Actors 1 and 2 in the diagram below illustrate the steps required for the JSON
+and Rust filters to be compiled into BPF, respectively.
 
-## Examples
+![seccompiler architecture](./docs/img/seccompiler_arch.jpg)
 
-TODO: Usage examples.
+### Filter definition
+
+Let us take a closer look at what a filter is composed of, and how it is
+defined:
+
+The smallest unit of the filter is the `SeccompCondition`, which is a
+comparison operation applied to the current system call. It’s parametrised by
+the argument index, the length of the argument, the operator and the actual
+expected value.
+
+Going one step further, a `SeccompRule` is a vector of `SeccompCondition`s,
+that must all match for the rule to be considered matched. In other words, a
+rule is a collection of **and-bound** conditions for a system call.
+
+Finally, at the top level, there’s the `SeccompFilter`. The filter can be
+viewed as a collection of syscall-associated rules, with a predefined on-match
+action and a default action that is returned if none of the rules match.
+
+In a filter, each system call number maps to a vector of **or-bound** rules.
+In order for the filter to match, it is enough that one rule associated to the
+system call matches. A system call may also map to an empty rule vector, which
+means that the system call will match, regardless of the actual arguments.
+
+The following diagram models a simple filter, that only allows `accept4`,
+`fcntl(any, F_SETFD, FD_CLOEXEC, ..)` and `fcntl(any, F_GETFD, ...)`.
+For any other system calls, the process will be killed.
+
+![filter diagram](./docs/img/seccomp_filter.png)
+
+As specified earlier, there are two ways of expressing the filters:
+
+1. JSON (documented in [json_format.md](docs/json_format.md));
+2. Rust code (documented by the library).
+
+See below examples of both representation methods, for a filter equivalent to
+the diagram above:
+
+#### Example JSON filter
+
+```json
+{
+    "main_thread": {
+        "default_action": "kill_process",
+        "filter_action": "allow",
```

Comment:
I am having difficulties understanding what is the difference between `default_action` and `filter_action`.